File: services/brain_api/src/brain_api/routers/ingest.py
"""Document ingestion and metadata management endpoints.

Service calls delegate to Qdrant/Neo4j/OpenAI clients which are all blocking
by design (the Python clients for these aren't natively async). We wrap each
call in `asyncio.to_thread` so long blocking I/O runs in the thread pool
instead of stalling the FastAPI event loop.
"""

from __future__ import annotations

import asyncio
import logging
from typing import Annotated, Any

from fastapi import APIRouter, Depends, HTTPException, status
from pydantic import BaseModel, Field

from brain_api.auth import require_api_key
from brain_api.services.ingest_service import IngestService
from brain_api.stores import Stores, get_stores

logger = logging.getLogger(__name__)
router = APIRouter()


class IngestRequest(BaseModel):
    tenant_id: str = Field(min_length=1, max_length=128)
    document_key: str = Field(min_length=1, max_length=255)
    title: str = Field(min_length=1, max_length=500)
    text: str = Field(min_length=1)
    tags: list[str] = Field(default_factory=list)
    access_keys: list[int] = Field(default_factory=list)
    use_knowledge_graph: bool = True
    metadata: dict[str, Any] = Field(default_factory=dict)


class RemoveDocumentRequest(BaseModel):
    tenant_id: str = Field(min_length=1, max_length=128)
    document_key: str = Field(min_length=1, max_length=255)


class UpdateMetadataRequest(BaseModel):
    tenant_id: str = Field(min_length=1, max_length=128)
    document_key: str = Field(min_length=1, max_length=255)
    title: str | None = Field(default=None, max_length=500)
    new_tags: list[str] | None = None
    new_access_keys: list[int] | None = None


class InitTenantRequest(BaseModel):
    tenant_id: str = Field(min_length=1, max_length=128)


def _get_service(stores: Annotated[Stores, Depends(get_stores)]) -> IngestService:
    return IngestService(stores)


@router.post("/ingest-document")
async def ingest_document(
    body: IngestRequest,
    service: Annotated[IngestService, Depends(_get_service)],
    _api_key: Annotated[str, Depends(require_api_key)],
) -> dict[str, Any]:
    """Ingest a document: chunk, embed, summarize, store in vector + graph."""
    try:
        return await asyncio.to_thread(
            service.ingest_document,
            tenant_id=body.tenant_id,
            document_key=body.document_key,
            title=body.title,
            text=body.text,
            tags=body.tags,
            access_keys=body.access_keys,
            use_knowledge_graph=body.use_knowledge_graph,
            metadata=body.metadata,
        )
    except Exception as e:
        logger.exception("Ingestion failed for %s", body.document_key)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Document ingestion failed",
        ) from e


@router.post("/remove-document")
async def remove_document(
    body: RemoveDocumentRequest,
    service: Annotated[IngestService, Depends(_get_service)],
    _api_key: Annotated[str, Depends(require_api_key)],
) -> dict[str, str]:
    await asyncio.to_thread(service.remove_document, body.tenant_id, body.document_key)
    return {"status": "deleted", "document_key": body.document_key}


@router.post("/update-document-metadata")
async def update_metadata(
    body: UpdateMetadataRequest,
    service: Annotated[IngestService, Depends(_get_service)],
    _api_key: Annotated[str, Depends(require_api_key)],
) -> dict[str, str]:
    await asyncio.to_thread(
        service.update_metadata,
        tenant_id=body.tenant_id,
        document_key=body.document_key,
        tags=body.new_tags,
        access_keys=body.new_access_keys,
        title=body.title,
    )
    return {"status": "updated"}


@router.post("/init-tenant")
async def init_tenant(
    body: InitTenantRequest,
    service: Annotated[IngestService, Depends(_get_service)],
    _api_key: Annotated[str, Depends(require_api_key)],
) -> dict[str, str]:
    try:
        await asyncio.to_thread(service.init_tenant, body.tenant_id)
    except Exception as e:
        logger.exception("Tenant init failed for %s", body.tenant_id)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Tenant initialization failed",
        ) from e
    return {"status": "initialized", "tenant_id": body.tenant_id}


@router.post("/remove-tenant")
async def remove_tenant(
    body: InitTenantRequest,
    service: Annotated[IngestService, Depends(_get_service)],
    _api_key: Annotated[str, Depends(require_api_key)],
) -> dict[str, str]:
    """Delete all Qdrant collections + Neo4j nodes for a tenant.

    Called by the API when an org is deleted. The service layer already
    logs per-store failures and proceeds, so the caller gets 200 even on
    partial failure — operators should monitor brain-api logs to spot
    cleanup gaps. (Failing loudly here would block PostgreSQL deletion
    upstream, which is usually worse than leaving orphan vectors.)
    """
    await asyncio.to_thread(service.remove_tenant, body.tenant_id)
    return {"status": "removed", "tenant_id": body.tenant_id}


@router.get("/documents/{tenant_id}/{document_key}/chunks")
async def get_document_chunks(
    tenant_id: str,
    document_key: str,
    stores: Annotated[Stores, Depends(get_stores)],
    _api_key: Annotated[str, Depends(require_api_key)],
    limit: int = 500,
) -> dict[str, Any]:
    """Return all chunks for a document, ordered by chunk_order."""
    try:
        chunks = await asyncio.to_thread(
            stores.vector.get_document_chunks, tenant_id, document_key, limit=limit
        )
    except Exception:
        logger.exception("Failed to fetch chunks for %s", document_key)
        raise HTTPException(
            status_code=status.HTTP_500_INTERNAL_SERVER_ERROR,
            detail="Failed to fetch chunks",
        )

    return {
        "document_key": document_key,
        "chunks": [
            {
                "id": c.id,
                "text": c.payload.get("text", ""),
                "chunk_order": c.payload.get("chunk_order", 0),
            }
            for c in chunks
        ],
    }

File: services/brain-api-go/internal/handlers/ingest.go
package handlers

import (
	"encoding/json"
	"log/slog"
	"net/http"

	"github.com/go-chi/chi/v5"

	"github.com/redarchlabs/red-arch-km-2/services/brain-api-go/internal/pipeline"
)

// IngestHandlers contains HTTP handlers for document ingestion.
type IngestHandlers struct {
	pipeline PipelineService
}

// NewIngestHandlers creates new ingest handlers.
func NewIngestHandlers(p PipelineService) *IngestHandlers {
	return &IngestHandlers{pipeline: p}
}

// IngestDocumentRequest is the request body for document ingestion.
type IngestDocumentRequest struct {
	TenantID          string         `json:"tenant_id"`
	DocumentKey       string         `json:"document_key"`
	Title             string         `json:"title"`
	Text              string         `json:"text"`
	Tags              []string       `json:"tags"`
	AccessKeys        []int          `json:"access_keys"`
	UseKnowledgeGraph bool           `json:"use_knowledge_graph"`
	Metadata          map[string]any `json:"metadata"`
}

// IngestDocument handles POST /ingest-document.
func (h *IngestHandlers) IngestDocument() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req IngestDocumentRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			jsonError(w, "invalid request body", http.StatusBadRequest)
			return
		}

		// Validate required fields
		if req.TenantID == "" {
			jsonError(w, "tenant_id is required", http.StatusBadRequest)
			return
		}
		if req.DocumentKey == "" {
			jsonError(w, "document_key is required", http.StatusBadRequest)
			return
		}
		if req.Title == "" {
			jsonError(w, "title is required", http.StatusBadRequest)
			return
		}
		if req.Text == "" {
			jsonError(w, "text is required", http.StatusBadRequest)
			return
		}

		result, err := h.pipeline.IngestDocument(r.Context(), pipeline.IngestRequest{
			TenantID:          req.TenantID,
			DocumentKey:       req.DocumentKey,
			Title:             req.Title,
			Text:              req.Text,
			Tags:              req.Tags,
			AccessKeys:        req.AccessKeys,
			UseKnowledgeGraph: req.UseKnowledgeGraph,
			Metadata:          req.Metadata,
		})
		if err != nil {
			slog.Error("ingestion failed", "document_key", req.DocumentKey, "error", err)
			jsonError(w, "document ingestion failed", http.StatusInternalServerError)
			return
		}

		jsonResponse(w, result, http.StatusOK)
	}
}

// RemoveDocumentRequest is the request body for document removal.
type RemoveDocumentRequest struct {
	TenantID    string `json:"tenant_id"`
	DocumentKey string `json:"document_key"`
}

// RemoveDocument handles POST /remove-document.
func (h *IngestHandlers) RemoveDocument() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req RemoveDocumentRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			jsonError(w, "invalid request body", http.StatusBadRequest)
			return
		}

		if req.TenantID == "" || req.DocumentKey == "" {
			jsonError(w, "tenant_id and document_key are required", http.StatusBadRequest)
			return
		}

		if err := h.pipeline.RemoveDocument(r.Context(), req.TenantID, req.DocumentKey); err != nil {
			slog.Error("remove document failed", "document_key", req.DocumentKey, "error", err)
			jsonError(w, "failed to remove document", http.StatusInternalServerError)
			return
		}

		jsonResponse(w, map[string]string{
			"status":       "deleted",
			"document_key": req.DocumentKey,
		}, http.StatusOK)
	}
}

// UpdateMetadataRequest is the request body for metadata updates.
type UpdateMetadataRequest struct {
	TenantID      string   `json:"tenant_id"`
	DocumentKey   string   `json:"document_key"`
	Title         *string  `json:"title,omitempty"`
	NewTags       []string `json:"new_tags,omitempty"`
	NewAccessKeys []int    `json:"new_access_keys,omitempty"`
}

// UpdateDocumentMetadata handles POST /update-document-metadata.
func (h *IngestHandlers) UpdateDocumentMetadata() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req UpdateMetadataRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			jsonError(w, "invalid request body", http.StatusBadRequest)
			return
		}

		if req.TenantID == "" || req.DocumentKey == "" {
			jsonError(w, "tenant_id and document_key are required", http.StatusBadRequest)
			return
		}

		if err := h.pipeline.UpdateMetadata(r.Context(), req.TenantID, req.DocumentKey, req.NewTags, req.NewAccessKeys, req.Title); err != nil {
			slog.Error("update metadata failed", "document_key", req.DocumentKey, "error", err)
			jsonError(w, "failed to update metadata", http.StatusInternalServerError)
			return
		}

		jsonResponse(w, map[string]string{"status": "updated"}, http.StatusOK)
	}
}

// TenantRequest is the request body for tenant operations.
type TenantRequest struct {
	TenantID string `json:"tenant_id"`
}

// InitTenant handles POST /init-tenant.
func (h *IngestHandlers) InitTenant() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req TenantRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			jsonError(w, "invalid request body", http.StatusBadRequest)
			return
		}

		if req.TenantID == "" {
			jsonError(w, "tenant_id is required", http.StatusBadRequest)
			return
		}

		if err := h.pipeline.InitTenant(r.Context(), req.TenantID); err != nil {
			slog.Error("init tenant failed", "tenant_id", req.TenantID, "error", err)
			jsonError(w, "tenant initialization failed", http.StatusInternalServerError)
			return
		}

		jsonResponse(w, map[string]string{
			"status":    "initialized",
			"tenant_id": req.TenantID,
		}, http.StatusOK)
	}
}

// RemoveTenant handles POST /remove-tenant.
func (h *IngestHandlers) RemoveTenant() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		var req TenantRequest
		if err := json.NewDecoder(r.Body).Decode(&req); err != nil {
			jsonError(w, "invalid request body", http.StatusBadRequest)
			return
		}

		if req.TenantID == "" {
			jsonError(w, "tenant_id is required", http.StatusBadRequest)
			return
		}

		if err := h.pipeline.RemoveTenant(r.Context(), req.TenantID); err != nil {
			slog.Error("remove tenant failed", "tenant_id", req.TenantID, "error", err)
			jsonError(w, "tenant removal failed", http.StatusInternalServerError)
			return
		}

		jsonResponse(w, map[string]string{
			"status":    "removed",
			"tenant_id": req.TenantID,
		}, http.StatusOK)
	}
}

// GetDocumentChunks handles GET /documents/{tenant}/{key}/chunks.
func (h *IngestHandlers) GetDocumentChunks() http.HandlerFunc {
	return func(w http.ResponseWriter, r *http.Request) {
		tenantID := chi.URLParam(r, "tenant")
		documentKey := chi.URLParam(r, "key")

		if tenantID == "" || documentKey == "" {
			jsonError(w, "tenant and document key are required", http.StatusBadRequest)
			return
		}

		limit := 500 // default
		if l := r.URL.Query().Get("limit"); l != "" {
			var lv int
			if _, err := parseIntParam(l, &lv); err == nil && lv > 0 && lv <= 1000 {
				limit = lv
			}
		}

		chunks, err := h.pipeline.GetDocumentChunks(r.Context(), tenantID, documentKey, limit)
		if err != nil {
			slog.Error("get chunks failed", "document_key", documentKey, "error", err)
			jsonError(w, "failed to fetch chunks", http.StatusInternalServerError)
			return
		}

		type chunkItem struct {
			ID         string `json:"id"`
			Text       string `json:"text"`
			ChunkOrder int    `json:"chunk_order"`
		}

		items := make([]chunkItem, len(chunks))
		for i, c := range chunks {
			order := 0
			if co, ok := c.Payload["chunk_order"].(float64); ok {
				order = int(co)
			}
			items[i] = chunkItem{
				ID:         c.ID,
				Text:       toString(c.Payload["text"]),
				ChunkOrder: order,
			}
		}

		jsonResponse(w, map[string]any{
			"document_key": documentKey,
			"chunks":       items,
		}, http.StatusOK)
	}
}

func toString(v any) string {
	if s, ok := v.(string); ok {
		return s
	}
	return ""
}

func parseIntParam(s string, out *int) (int, error) {
	var v int
	_, err := parseIntFromString(s, &v)
	if err != nil {
		return 0, err
	}
	*out = v
	return v, nil
}

func parseIntFromString(s string, out *int) (int, error) {
	var v int
	n, err := json.Number(s).Int64()
	if err != nil {
		return 0, err
	}
	v = int(n)
	*out = v
	return v, nil
}
